### PR TITLE
adding graph relabeling and sublattice mappings.

### PR DESCRIPTION
**Pull request by boothby**
I'm making two significant additions in this PR, and I don't expect it to be merged as-is.  Happy to make them separate issues (also todo; documentation is incomplete, there are no tests, ~I haven't even checked for obvious typos/omitted imports~).  Both are additions to the `pegasus_coordinates` class.  My plan is to repeat this work for Chimera and Zephyr, but I'm pausing here to get some feedback on the interface, and what I'm calling things.  **Yes please bikeshed this**

**Graph Converters:** I've added three functions to convert graphs from one labeling to another.  For example, if you've made a `pegasus_graph` `p` with integer labels, but you'd prefer to work with coordinate labels, this saves on boilerplate.  What was previously
```python
p = pegasus_graph(m, ...)
coords = pegasus_coordinates(m)
nodes = coords.iter_linear_to_pegasus(p.nodes())
edges = coords.iter_linear_to_pegasus_pairs(p.edges())
p_coord = pegasus_graph(m, node_list = nodes, edge_list = edges)
```
is now
```python
p = pegasus_graph(m, ...)
coords = pegasus_coordinates(m)
p_coord = coords.graph_to_pegasus(p)
```

**Sublattice Mappings:** This adds functionality to make homomorphisms from small Pegasus graphs into large Pegasus graphs.  This is broken into two steps: (step 1) generating "offsets" which are used to produce (step 2) mappings from a `pegasus_graph(m_small)` to `pegasus_graph(m_large)`.  We show a barely-contrived use case below; searching for fully-yielded P3 subgraphs:

```python
host_graph = pegasus_graph(m, ...)
host_coords = pegasus_coordinates(m)
mask_graph = pegasus_graph(3)
for offset in host_coords.sublattice_offsets(3):
    mapping = host_coords.sublattice_linear_mapping(3, offset)
    subgraph = host_graph.subgraph(mapping(q) for q in mask_graph)
    if subgraph.number_of_edges() == mask_graph.number_of_edges():
       print("found a fully-yielded P3 graph with offset", offset)
```

### CODE REVIEW

**Comment by boothby** [comment]
@pau557 @hhtong @jackraymond I think you'll interact with the sublattice mappings the most; @arcondello @randomir I expect you'll have opinions about that and the graph relabeling interface.

**Comment by boothby** [comment]
I've been asked to do `nice_coordinates` for Zephyr, and since I already had this branch going and my head was already deep in coordinate spaces, added that to the scope of this PR and finished the whole thing off.  I'm still open to bikeshedding, as long as it's roughly search-and-replace-able.

File: dwave_networkx/generators/pegasus.py [hunk 1]
```diff
@@ -256,24 +260,39 @@ def efilter(e): return qfilter(*e[0]) and qfilter(*e[1])
...
     if data:
         v = 0
+        def coord_label():
+            return q
+        def int_label():
+            return v
+        if nice_coordinates:
+            def this_label():
+                return pegasus_to_nice(q)
+            other_name = 'linear_index'
```
Comment (jackraymond) [inline dwave_networkx/generators/pegasus.py:270]: nice_index?

File: dwave_networkx/generators/zephyr.py [hunk 2]
```diff
@@ -57,25 +57,39 @@ def zephyr_graph(m, t=4, create_using=None, node_list=None, edge_list=None,
...
+        if ``coordinates`` is True, then both attributes are present.
     coordinates : bool, optional (default False)
-        If True, node labels are 5-tuple Zephyr indices.
+        If True, node labels are 5-tuple Zephyr indices.  Ignored if the
+        `nice_coordinates` parameter is True.
+    nice_coordinates: bool, optional (default False)
+        The graph uses a "nicer" coordinate system, more compatible with Chimera
+        addressing.  These coordinates are 5-tuples taking the form 
+        :math:`(j, y, x, u, k)` where :math:`0 <= j < 2`, :math:`0 <= x < m`,
+        :math:`0 <= y < m`, :math:`0 <= u < 2`, and :math:`0 <= k < 4`.
```
Comment (jackraymond) [inline dwave_networkx/generators/zephyr.py:68]: 4 -> 2t

File: dwave_networkx/generators/pegasus.py (hunk 1 above)
Comment (boothby) [reply dwave_networkx/generators/pegasus.py:270]: Sorry, yeah, this code is quite annoying.  I tried to homogenize the approach behind filling in the labels... I think I can do better.

Comment (boothby) [reply dwave_networkx/generators/pegasus.py:270]: But, no, not nice_index -- `other_name` corresponds to the name of the field that's being filled in by `other_label`

**Comment by jackraymond** [comment]
In your commit comment you mention graph_to_pegasus, but that function seems to be absent.

File: dwave_networkx/generators/pegasus.py [hunk 3]
```diff
@@ -744,6 +772,140 @@ def iter_nice_to_linear_pairs(self, nlist):
...
+            node_list=nodes,
+            edge_list=edges,
+            data=g.graph['data'],
+            offset_lists=(
+                g.graph["vertical_offsets"],
+                g.graph["horizontal_offsets"],
+            ),
+        )
+
+    def graph_to_pegasus(self, g):
```
Comment (boothby) [inline dwave_networkx/generators/pegasus.py:802]: @jackraymond

**Comment by jackraymond** [comment]
A couple of superficial problems with nice coordinates, might be worth thinking about. How nice do we want nice to be? 
 
Having negative coordinate values in the nice scheme is a little bit awkward to me. This could be addressed by a modification of the cell definition (i.e. having a few x, y values of m+1, rather than a few x,y values of -1 - pushing awkward boundary conditions to high integers rather than near the origin). 
Some documentation describes nice_coordinates x and y as limited to the range [0,m), which is inaccurate without greater context.

The nice coordinate scheme indices don't play nice with the visualization tools (draw_zephyr). Variables ordered in x (or y) are not ordered vertically and horizontally in standard visualization.
Lots of other nice features that could be ruined in trying to workaround these two issues of course.. 
Z(m=3,t=1) shows features.
```
m = 2
t = 1
graph = dnx.zephyr_graph(m=m,t=t)
graph = coords.graph_to_nice(graph)
dnx.draw_zephyr(graph,with_labels=True)
```

**Comment by boothby** [comment]
@jackraymond and I had a sidebar; the outcome is to drop the current proposal for `nice_coordinates` and use a 4-coordinate Chimera addressing scheme as depicted below (a C_2n embeds directly in to a Z_n, mapping external couplers to odd couplers).  I'll take a crack at implementing the interface tomorrow.

foo

**Review by pau557 — APPROVED** [review]
LGTM. I appreciate the sublattice work

File: dwave_networkx/generators/chimera.py [hunk 4]
```diff
@@ -449,6 +462,97 @@ def iter_linear_to_chimera_pairs(self, plist):
         """
         return self._pair_repack(self.iter_linear_to_chimera, plist)
 
+    def graph_to_linear(self, g):
+        """Return a copy of the graph g relabeled to have linear indices"""
+        labels = g.graph.get('labels')
+        if labels == 'int':
```
Comment (pau557) [inline dwave_networkx/generators/chimera.py:468]: Do we have two alternative nomenclatures with `linear` and `int` coordinates?

Comment (boothby) [reply dwave_networkx/generators/chimera.py:468]: Yes, I'm following the convention set in `chimera_graph`

Comment (pau557) [reply dwave_networkx/generators/chimera.py:468]: Ok, so `int` just gets used inside G.graph[`labels`] for Chimera, Pegasus, Zephyr. Outside of there we call them `linear`

**Comment by boothby** [comment]
After time to digest, I came up with a much simpler interface for sublattice mappings.  On a D-Wave internal slack channel, this proved to be a popular option.

```python
def pegasus_sublattice_mappings(source, target):
    """yields mappings from Chimera or Pegasus graphs into Pegasus graphs"""

def chimera_sublattice_mappings(source, target):
    """yields mappings from Chimera graphs into Chimera graphs"""
    
def zephyr_sublattice_mappings(source, target):
    """yields mappings from Chimera or Zephyr graphs into Zephyr graphs"""
```

Unable to decide between my original definition of `nice_coordinates` and the one(s) proposed by Jack, I decided that neither deserves the name "nice".  Instead, I've made it easy, through the interface sketched above, to simply generate all sublattice mappings from `chimera_graph(m, m, 8)` or  `chimera_graph(2*m, 2*m, 4)` to `zephyr_graph(m, 8)` through the same unified interface:

```python
import dwave_networkx as dnx
from matplotlib import pyplot as plt
c2 = dnx.chimera_graph(2)
z2 = dnx.zephyr_graph(2)
for f in dnx.zephyr_sublattice_mappings(c2, z2):
    plt.figure(figsize=(10,10))
    dnx.draw_zephyr_embedding(z2, {0:[f(v) for v in c2]})
    plt.savefig(f"single{''.join(map(str,f.offset))}.png")
    plt.close()
c28 = dnx.chimera_graph(2, t=8)
z4 = dnx.zephyr_graph(4)
for f in dnx.zephyr_sublattice_mappings(c28, z4):
    plt.figure(figsize=(10,10))
    dnx.draw_zephyr_embedding(z4, {0:[f(v) for v in c28]})
    plt.savefig(f"double{''.join(map(str,f.offset))}.png")
    plt.close()
```

single
double

File: dwave_networkx/drawing/zephyr_layout.py [hunk 5]
```diff
@@ -79,7 +78,10 @@ def zephyr_layout(G, scale=1., center=None, dim=2):
         m = G.graph.get('rows')
         t = G.graph.get('tile')
         coord = zephyr_coordinates(m, t)
-        pos = {v: xy_coords(*coord.linear_to_zephyr(v)) for v in G.nodes()}
+        if G.graph.get('labels') == 'nice':
```
Comment (boothby) [inline dwave_networkx/drawing/zephyr_layout.py:81]: oops

**Comment by boothby** [comment]
@arcondello I've rebased & rearranged the deltas to be more cogent

**Comment by arcondello** [comment]
@boothby , can you rebase one more time to fix the CI issues?

**Review by arcondello — APPROVED** [review]

**Comment by arcondello** [comment]
@jackraymond , are you finished reviewing?

**Pull request merged**